tools/mosread.1
.\"
.TH "MOSREAD" "1" "January 2025" "mosread" "User Commands"
.SH "NAME"
mosread \- SAMOS reader
.SH "SYNOPSIS"
.B mosread
[\fIoptions\fR] \fIdevice mode \fR[\fInode\fR] [\fIpath\fR]
.SH "DESCRIPTION"
Tool for a read\-only access on Smaky SAMOS.
.TP
\fB\-h\fR \fB\-\-help\fR
Display help message.
.TP
\fB\-v\fR \fB\-\-version\fR
Show version.
.TP
\fB\-l\fR \fB\-\-mos\-logger\fR
Turn ON the MOS logger.
.TP
\fBdevice\fR
file.di for disk image
.br
/dev/fd0 for floppy disk
.br
/dev/sda for hard disk, etc, ...
.TP
\fBmode\fR
\fBlist\fR : list the content of a node
.br
\fBget\fR  : copy a file (or a directory) from the Smaky disk in a local directory
.TP
\fBnode\fR
The tree with the file (or directory) for \fBget\fR or \fBlist\fR.
.br
example: foo/bar/toto.sr
.TP
\fBpath\fR
You can specify a path to save the file (with \fBget\fR mode).
.SH "AUTHOR"
Written by Mathieu Schroeter <mathieu@schroetersa.ch>.
.SH "REPORTING BUGS"
Report bugs to <\fImathieu@schroetersa.ch\fP>.
.SH "COPYRIGHT"
Copyright \(co 2025 Mathieu Schroeter

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


ChangeLog
2025-MM-DD  Mathieu Schroeter <mathieu@schroetersa.ch>

	* Release 2.0.0

	* The Windows version supports binary files (dumps, like with Linux)
	  and not just physical devices.

	* fosmount: replace the -i and -j options by only the -i option with the
	  major difference that BMP files are produced for .IMAGE and .COLOR
	  files instead of PBM and XPM2 which are not supported on Windows.

	* fosmount: add a new -t, --text option useful to convert some text
	  files to the ISO-8859-1 charset.

	* fosread: add recursive copy of a whole directory with the "get" mode.

	* fosread: add -i, --image-bmp and -t, --text options to the fosread
	  command. It enables the possibility to convert these files on Windows
	  like in the case of fosmount for Linux.

	* mosread: now tool to read and extract content of SAMOS disks and
	  images.

2024-10-08  Mathieu Schroeter <mathieu@schroetersa.ch>

	* Release 1.0.1

	* fosmount: move to fuse3. Thanks to Jakob Haufe from Debian.

2023-03-26  Mathieu Schroeter <mathieu@schroetersa.ch>

	* Release 1.0.0

	* fosdd: new tool that can be used in order to convert floppy device
	  or image into hard disk image and vice-versa.

	* fosmount: the PBM and XPM images use a shorter filename.

2010-03-14  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Release 0.4.0

	* Many improvements in the build scripts and Makefiles. The build for
	  Windows uses the --cross-compile option with the configure script.
	  The external scripts to build for Windows, provided with the older
	  versions are no longer necessary. Now the DLLs are created in the
	  right way.
	  The pkg-config files are now available for libfosfat and libfosgra.

	* libfosgra: add support for the decoding of .COLOR files.

	* fosmount: add the conversion on the fly of .COLOR files to .XPM files.
	  Note that the XPM files use the version 2 of X PixMap. It needs a real
	  application like The Gimp in order to read the files. Or the tool SXPM
	  from the X11 (X.Org) project.

	* smascii: add a manpage. Thanks to Didier Raboud from Debian.

	* libw32disk: this library doesn't rely anymore to C++. It is now fully
	  rewritten in C with a proper license in order to be redistribuable by
	  Debian.

2009-07-07  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New beta released 0.3.2

	* libfosgra: new library to support .IMAGE decoding.

	* fosmount: some improvements to convert on the fly the images [.IMAGE]
	  to PBM (portable bitmap format) with libfosgra.

	* Fix doxygen, packaging and Makefiles.

2009-07-04  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New beta released 0.3.1

2009-05-06  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* fosrec: Add a new tool to easily recovery files.

	* Fix Debian/Ubuntu packaging.

2009-01-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* A lot of cleaning and all doxygen documentations are only in the
	  public header file.

	* Add option in configure to generate doxygen HTML and PDF outputs.

2008-02-04  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Improve a lot the code lisibility.

	* Fix some memleaks when errors appear.

2008-01-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Add manpages for fosmount and fosread.

2007-12-16  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New major beta released 0.3.0

	* Add Window$ support, and use MinGW32 to compile DLL libraries,
	  fosread and smascii. A script package-win32.sh can be used to create
	  a zip package.

	* libfosfat: add support for Window$ through libw32disk.

	* libw32disk: add C wrapper to use original Disk.cxx class with
	  libfosfat.

	* fosread: fix symlink support, now that is possible to read like a
	  directory. Add support to list and restore deleted files.

2007-12-13  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: first implementation of an undelete tool just disabled by
	  default. The next step is added some options to list and restore
	  files.

2007-12-12  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: add support for multiple devices. A new main structure is
	  used instead of global variables. Only the logger is always global.

2007-12-09  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Fourth beta released 0.2.3

	* Some fixes, cleaned code, ...

2007-07-13  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Change license GPLv2 to GPLv3.

	* Fix a lot of memory leaks.

2007-07-07  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Third beta released 0.2.2

2007-06-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: change the auto detection function for a better way. Use
	  the name was very bad because a lot of disks have no name. Then the
	  test is does with the CHK (look the code for more informations).

2007-06-24  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Add auto detection for the disk type. The type can be forced with
	  fosmount and fosread if necessary.

	* libfosfat: an internal logger can be enabled to know what the FOS is
	  doing. A new option is available in fosread and fosmount.

2007-06-22  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Second beta released 0.2.1

	* Add support for symbolic links. The Smaky's soft links are now
	  converted to symlink.

	* libfosfat: the cache is now released when the device is closed.

2007-06-20  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: a new system of caching is implemented to improve the
	  speed. Now that will be very fast for looking in a directory. The
	  system must be fixed for that the memory is released.

	* fosmount: a bug is fixed for that FUSE is always loaded as
	  single-thread. If not there is a really a big problem when data are
	  loaded with libfosfat.

2007-05-10  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* First beta released 0.2.0

2007-04-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>


tools/mosread.c
/*
 * SAMOS mosread: tool in read-only for Smaky file system
 * Copyright (C) 2025 Mathieu Schroeter <mathieu@schroetersa.ch>
 *
 * This file is part of Fosfat.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h> /* strcmp strcpy */
#include <getopt.h>
#include <sys/stat.h> /* mkdir */

#include "fosfat.h"
#include "fosfat_internal.h"


#undef HELP_DEVICE
#define HELP_DEVICE \
"file.di  : disk image\n" \
"                       /dev/fd0 : floppy disk\n" \
"                       /dev/sda : hard disk, etc\n"

#define HELP_TEXT \
"Tool for a read-only access on a Smaky MOS disk. mosfat-" LIBFOSFAT_VERSION_STR "\n\n" \
"Usage: mosread [options] device mode [node] [path]\n\n" \
" -h --help             this help\n" \
" -v --version          version\n" \
" -l --mos-logger       that will turn on the MOS logger\n\n" \
" device                " HELP_DEVICE \
" mode\n" \
"   list                list the content of a node\n" \
"   get                 copy a file (or a directory) from the Smaky's disk in\n" \
"                       a local directory\n" \
" node                  the tree with the file (or folder)" \
" for 'get' or 'list'\n" \
"                       example: foo.dr/bar.dr/toto.tx\n" \
" path                  you can specify a path to save" \
" the file (with get mode)\n" \
"\nPlease, report bugs to <mathieu@schroetersa.ch>.\n"

#define VERSION_TEXT "mosread-" LIBFOSFAT_VERSION_STR "\n"


static inline int
my_mkdir (const char *pathname)
{
#ifdef _WIN32
  return mkdir (pathname);
#else
  return mkdir (pathname, 0755);
#endif /* !_WIN32 */
}

/*
 * Print date and hour.
 *
 * time date and hour
 */
static void
print_date (mosfat_time_t *time)
{
  printf (" %02u-%02u-%02u", time->year, time->month, time->day);
}

/*
 * Print a file in the list.
 *
 * file description
 */
static void
print_file (mosfat_file_t *file)
{
  char filename[MOSFAT_NAMELGT];

  strncpy (filename, file->name, MOSFAT_NAMELGT - 1);
  filename[MOSFAT_NAMELGT - 1] = '\0';

  printf ("%c %8i", file->att.isdir ? 'd' : '-', file->size);
  print_date (&file->time);
  printf (" %s", filename);
  printf ("\n");
}

/*
 * List the content of a directory.
 *
 * mosfat       handle
 * path         where in the tree
 * return true if it is ok
 */
static int
list_dir (mosfat_t *mosfat, const char *loc)
{
  char *path;
  mosfat_file_t *files, *first_file;

  path = strdup (loc);

  if ((files = mosfat_list_dir (mosfat, path)))
  {
    first_file = files;
    printf ("path: %s\n\n", path);
    printf ("      size date     filename\n");
    printf ("      ---- -------- --------\n");

    do
    {
      print_file (files);
    } while ((files = files->next_file));

    printf ("\nd:directory\n");
    mosfat_free_listdir (first_file);
  }
  else
  {
    free (path);
    fprintf (stderr, "ERROR: I can't found this path!\n");
    return 0;
  }

  free (path);

  return 1;
}

/*
 * Copy a file from the disk.
 *
 * mosfat       handle
 * path         where in the tree
 * dst          where in local
 * return true if it is ok
 */
static int
get_file (mosfat_t *mosfat, const char *path, const char *dst)
{
  int res = 0;
  char *new_file, *name = NULL;

  if (!strcasecmp (dst, "./"))
    new_file = strdup ((strrchr (path, '/') ? strrchr (path, '/') + 1 : path));
  else
    new_file = strdup (dst);

  if (!mosfat_isdir (mosfat, path))
  {
    printf ("File \"%s\" is copying ...\n", path);
    name = strdup (new_file);

    if (mosfat_get_file (mosfat, path, name))
    {
      res = 1;
      printf ("Okay..\n");
    }
    if (res == 0)
      fprintf (stderr, "ERROR: I can't copy the file!\n");
  }
  else
    fprintf (stderr, "ERROR: I can't copy a directory or a link!\n");

  free (new_file);
  if (name)
    free (name);
  return res;
}

static int
get_dir (mosfat_t *mosfat, const char *loc, const char *dst)
{
  char *path;
  mosfat_file_t *file, *first_file;

  path = strdup (loc);

  if ((file = mosfat_list_dir (mosfat, path)))
  {
    char out[4096] = {0};
    char in[256]  = {0};
    first_file = file;

    do
    {
      snprintf (in , sizeof (in),  "%s/%s", loc, file->name);
      snprintf (out, sizeof (out), "%s/%s", dst, file->name);
      remove_dup_slashes (in);
      remove_dup_slashes (out);

      if (file->att.isdir)
      {
        char *it = strrchr (out, '.');
        if (it)
          *it = '\0'; /* drop .DR from the name */
        my_mkdir (out);
        get_dir (mosfat, in, out);
      }
      else
      {
        if (file->size > 0)
          get_file (mosfat, in, out);
        else
          fprintf (stderr, "WARNING: skip empty file (0 bytes)\n");
      }
    } while ((file = file->next_file));

    mosfat_free_listdir (first_file);
  }
  else
  {
    free (path);
    fprintf (stderr, "ERROR: I can't found this path!\n");
    return 0;
  }

  free (path);

  return 1;
}

/* Print help. */
static void
print_info (void)
{
  printf (HELP_TEXT);
}

/* Print version. */
static void
print_version (void)
{
  printf (VERSION_TEXT);
}

int
main (int argc, char **argv)
{
  int res = 0, i, next_option;
  char *device = NULL, *mode = NULL, *node = NULL, *path = NULL;
  mosfat_t *mosfat;

  const char *const short_options = "hlv";

  const struct option long_options[] = {
    { "help",         no_argument, NULL, 'h' },
    { "mos-logger",   no_argument, NULL, 'l' },
    { "version",      no_argument, NULL, 'v' },
    { NULL,           0,           NULL,  0  }
  };

  /* check options */
  do
  {
    next_option = getopt_long (argc, argv, short_options, long_options, NULL);
    switch (next_option)
    {
    default :           /* unknown */
    case '?':           /* invalid option */
    case 'h':           /* -h or --help */
      print_info ();
      return -1;
    case 'v':           /* -v or --version */
      print_version ();
      return -1;
    case 'l':           /* -l or --mos-logger */
      fosfat_logger (1);
      break ;
    case -1:            /* end */
      break ;
    }
  } while (next_option != -1);

  if (argc < optind + 2)
  {
    print_info ();
    return -1;
  }

  for (i = optind; i < argc; i++)
  {
    if (i == optind)
      device = strdup (argv[optind]);
    else if (i == optind + 1)
      mode = strdup (argv[optind + 1]);
    else if (i == optind + 2)
      node = strdup (argv[optind + 2]);
    else if (i == optind + 3)
      path = strdup (argv[optind + 3]);
  }

  /* Open the image disk */
  if (!(mosfat = mosfat_open (device)))
  {
    fprintf (stderr, "Could not open %s for reading!\n", device);
    res = -1;
  }

  /* Get globals informations on the disk */
  if (!res)
  {
    /* Show the list of a directory */
    if (!strcasecmp (mode, "list"))
    {
      if (!node)
      {
        if (!list_dir (mosfat, "/"))
          res = -1;
      }
      else if (node)
      {
        if (!list_dir (mosfat, node))
          res = -1;
        free (node);
      }
    }

    /* Get a file from the disk */
    else if (!strcmp (mode, "get") && node)
    {
      if (mosfat_isdir (mosfat, node))
        get_dir (mosfat, node, path ? path : "./");
      else
        get_file (mosfat, node, path ? path : "./");
      free (node);
      free (path);
    }
    else
      print_info ();
  }

  /* Close the disk */
  mosfat_close (mosfat);
  free (device);
  free (mode);

  return res;
}
